File: src/services/media.ts
import { invoke } from "@tauri-apps/api/core";

export interface Media {
  journal_id: string;
  customenabled: boolean;
  backgroundimage?: string;
  isbgenabled?: boolean;
  primary_color?: string;
  secondary_color?: string;
  text_color?: string;
  song?: string;
  font_title?: string;
  font_body?: string;
}

// Create a new media record
export async function createMedia(media: Media): Promise<void> {
  try {
    await invoke("create_media", {
      journalId: media.journal_id,
      customenabled: media.customenabled,
      backgroundimage: media.backgroundimage,
      isbgenabled: media.isbgenabled,
      primaryColor: media.primary_color,
      secondaryColor: media.secondary_color,
      textColor: media.text_color,
      song: media.song,
      fontTitle: media.font_title,
      fontBody: media.font_body,
    });
    console.log("Media created successfully");
  } catch (error) {
    console.error("Failed to create media:", error);
    throw error;
  }
}

// Update an existing media record
export async function updateMedia(media: Media): Promise<void> {
  try {
    await invoke("update_media", {
      journalId: media.journal_id,
      customenabled: media.customenabled,
      backgroundimage: media.backgroundimage,
      isbgenabled: media.isbgenabled,
      primaryColor: media.primary_color,
      secondaryColor: media.secondary_color,
      textColor: media.text_color,
      song: media.song,
      fontTitle: media.font_title,
      fontBody: media.font_body,
    });
    console.log("Media updated successfully");
  } catch (error) {
    console.error("Failed to update media:", error);
    throw error;
  }
}

// Delete a media record by journal_id
export async function deleteMedia(journalId: string): Promise<void> {
  try {
    await invoke("delete_media", { journalId });
    console.log("Media deleted successfully");
  } catch (error) {
    console.error("Failed to delete media:", error);
    throw error;
  }
}

// Get a single media record by journal_id
export async function getMedia(journalId: string): Promise<Media | null> {
  try {
    const media = await invoke<Media | null>("get_media", { journalId });
    console.log("Media fetched successfully:", media);
    return media;
  } catch (error) {
    console.error("Failed to fetch media:", error);
    throw error;
  }
}

// Get all media records
export async function getAllMedia(): Promise<Media[]> {
  try {
    const mediaList = await invoke<Media[]>("get_all_media");
    console.log("Media list fetched successfully:", mediaList);
    return mediaList;
  } catch (error) {
    console.error("Failed to fetch media list:", error);
    throw error;
  }
}

File: src-tauri/src/commands/media.rs
use tauri::async_runtime::Mutex;
use tauri::{AppHandle, Emitter, Manager, Window};
use crate::services::db::DatabaseManager;
use crate::services::media;
use crate::models;
use anyhow::Result;
use fastrand::i8;

#[tauri::command]
pub async fn create_media(
    app: tauri::AppHandle,
    window: tauri::Window,
    journal_id: String,
    customenabled: bool,
    backgroundimage: Option<String>,
    isbgenabled: Option<bool>,
    primary_color: Option<String>,
    secondary_color: Option<String>,
    text_color: Option<String>,
    song: Option<String>,
    font_title: Option<String>,
    font_body: Option<String>,
) -> Result<(), String> {
    let db_st = app.state::<Mutex<DatabaseManager>>();
    let db_manager = db_st.lock().await;
    let conn = db_manager.get_connection().map_err(|e| e.to_string())?;

    let wall_id: i8 = fastrand::i8(..=7);
    // Apply default values for optional parameters
    let default_backgroundimage = Some(format!("/assets/{wall_id}.avif"));
    let default_primary_color = Some("#FFFFFF".to_string());
    let default_secondary_color = Some("#000000".to_string());
    let default_text_color = Some("#333333".to_string());
    let default_song = Some("default_song.mp3".to_string());
    let default_font_title = Some("Playfair Display".to_string());
    let default_font_body = Some("Lexend Deca".to_string());

    let result = media::create_media(
        &conn,
        journal_id,
        customenabled,
        backgroundimage.or(default_backgroundimage),
        isbgenabled.unwrap_or(false), // Default to false if not provided
        primary_color.or(default_primary_color),
        secondary_color.or(default_secondary_color),
        text_color.or(default_text_color),
        song.or(default_song),
        font_title.or(default_font_title),
        font_body.or(default_font_body),
    )
    .map_err(|e| e.to_string());

    if result.is_ok() {
        window.emit("db:media-created", ()).map_err(|e| e.to_string())?;
    } else {
        window
            .emit("db:error", "Failed to create media")
            .map_err(|e| e.to_string())?;
    }

    result
}

#[tauri::command]
pub async fn update_media(
    app: AppHandle,
    window: Window,
    journal_id: String,
    customenabled: bool,
    backgroundimage: Option<String>,
    isbgenabled: Option<bool>,
    primary_color: Option<String>,
    secondary_color: Option<String>,
    text_color: Option<String>,
    song: Option<String>,
    font_title: Option<String>,
    font_body: Option<String>,
) -> Result<(), String> {
    let db_st = app.state::<Mutex<DatabaseManager>>();
    let db_manager = db_st.lock().await;
    let conn = db_manager.get_connection().map_err(|e| e.to_string())?;

    let result = media::update_media(
        &conn,
        journal_id,
        customenabled,
        backgroundimage,
        isbgenabled,
        primary_color,
        secondary_color,
        text_color,
        song,
        font_title,
        font_body,
    )
    .map_err(|e| e.to_string());

    if result.is_ok() {
        window.emit("db:media-updated", ()).map_err(|e| e.to_string())?;
    } else {
        window
            .emit("db:error", "Failed to update media")
            .map_err(|e| e.to_string())?;
    }

    result
}

#[tauri::command]
pub async fn delete_media(
    app: AppHandle,
    window: Window,
    journal_id: String,
) -> Result<(), String> {
    let db_st = app.state::<Mutex<DatabaseManager>>();
    let db_manager = db_st.lock().await;
    let conn = db_manager.get_connection().map_err(|e| e.to_string())?;

    let result = media::delete_media(&conn, journal_id).map_err(|e| e.to_string());

    if result.is_ok() {
        window.emit("db:media-deleted", ()).map_err(|e| e.to_string())?;
    } else {
        window
            .emit("db:error", "Failed to delete media")
            .map_err(|e| e.to_string())?;
    }

    result
}

#[tauri::command]
pub async fn get_media(
    app: AppHandle,
    window: Window,
    journal_id: String,
) -> Result<Option<models::Media>, String> {
    let db_st = app.state::<Mutex<DatabaseManager>>();
    let db_manager = db_st.lock().await;
    let conn = db_manager.get_connection().map_err(|e| e.to_string())?;

    let result = media::get_media(&conn, journal_id).map_err(|e| e.to_string());

    if result.is_ok() {
        window.emit("db:media-fetched", ()).map_err(|e| e.to_string())?;
    } else {
        window
            .emit("db:error", "Failed to fetch media")
            .map_err(|e| e.to_string())?;
    }

    result
}

#[tauri::command]
pub async fn get_all_media(
    app: AppHandle,
    window: Window,
) -> Result<Vec<models::Media>, String> {
    let db_st = app.state::<Mutex<DatabaseManager>>();
    let db_manager = db_st.lock().await;
    let conn = db_manager.get_connection().map_err(|e| e.to_string())?;

    let result = media::get_all_media(&conn).map_err(|e| e.to_string());

    if result.is_ok() {
        window.emit("db:media-list-fetched", ()).map_err(|e| e.to_string())?;
    } else {
        window
            .emit("db:error", "Failed to fetch media list")
            .map_err(|e| e.to_string())?;
    }

    result
}